### DISCUSSION TITLE
Openproject notification

### DISCUSSION THREAD

**Discussion by wolfjudge** [girish17/op-mattermost#36] [Q&A] [answered]
Hello,
Just wanted to check if there is a possibility to have this integrated in a way that openproject send notifications to specific mattermost channel?
Tnx in advance 
Judge

**Comment by girish17** [comment, marked as answer]
OpenProject notifications can be configured (screenshot attached) through its [Webhooks](https://www.openproject.org/docs/system-admin-guide/incoming-and-outgoing/webhooks/) and the Payload URL can be pointed to op-mattermost RESTful POST endpoint running on Node JS server. The code snippets for the same are as follows:
1. Add a route in resource/routes.js like this -
```
 app.post('/notifyChannel', (req, res) => {
    console.log("Notify channel request: ", JSON.stringify(req.body, null, 2));
    uiActions.notifyChannel(req, res, axios);
  });
```
2. In resource/uiActions.js handle the Payload and extract the contents such as action, comment, project name, author etc. Here only action type is notified -
```
 notifyChannel(req, res, axios) {
    let msg = req.body.action;
    console.log("Notification message: ", msg);
    this.message.showNotification(res, axios, msg);
  }
```
3. Create an ephemeral message to the Mattermost channel already used by the slash command - 
```
showNotification(res, axios, msg) {
	    console.log("Creating notification for ChannelID: ", this.channel_id);
	    this.msgObj.post.message = msg;
	    console.log("Notification post message: ", this.msgObj);
            axios.post(this.mmURL + 'posts/ephemeral',
                this.msgObj, this.config).then((result) => {
                if (result.data) {
                    res.send('Show notification post succeeded!').status(200);

                }
                else {
                    console.log('Show notification post failed!');
                    res.send().status(400);

                }
            }).catch((err) => {
                console.log('Show notification post failed: %o', err);
                res.send().status(500);
            });
    }
```

The complete code commits can be found [here](https://github.com/girish17/op-mattermost/commit/e3f78025eabe7da4c13136b3fcf4c8850bcd6c10) and [here](https://github.com/girish17/op-mattermost/commit/3f2cd2a1439e889d7967d607774d81507df5c650#diff-b69868a08f4defe5e53a323bef97a0c38525bb9ffd6fd633ccc7772d0d68f3f7).

![webhookOP](https://user-images.githubusercontent.com/25923137/184819169-b33cd4b4-74ba-4055-88bf-299cc024111d.png)
.
